src/modules/Selectors.jl

```diff
@@ -0,0 +1,151 @@
+        Selectors.selector(EvenOdd, xs, x)
+    end
+    foreach(println, xs)
+end
+```
+
+## Why is this useful?
+
+`Selectors` allows for the "matching" and "execution" code to remain isolated from each
+other. Additional selector cases can be user-defined and hook into an already-defined
+selector. For example, we can insert another case into the `EvenOdd` selector defined above
+with
+
+```julia
+abstract Zero <: EvenOdd
+
+# Insert this case prior to the `Even` case.
+Selectors.order(::Type{Zero}) = 0.0
+
+# We match this case when `x` is `0`.
+Selectors.matcher(::Type{Zero}, xs, x) = x === 0
+
+# The `Zero` case empties the vector `xs`.
+Selectors.runner(::Type{Zero}, xs, x) = empty!(xs)
+```
+
+Note that once a `selector` method has been compiled for a specific type such as `EvenOdd`
+then adding more cases will not have any effect.
+"""
+module Selectors
+
+using Compat
+
+"""
+Root selector type. Each user-defined selector must subtype from this, i.e.
+
+```julia
+abstract MySelector <: Selectors.AbstractSelector
+
+abstract First  <: MySelector
+abstract Second <: MySelector
+```
+"""
+abstract AbstractSelector
+
+"""
+Define the precedence of each case in a selector, i.e.
+
+```julia
+Selectors.order(::Type{First})  = 1.0
+Selectors.order(::Type{Second}) = 2.0
+```
+
+Note that the return type must be `Float64`. Defining multiple case types to have the same
+order will result in undefined behaviour.
+"""
+function order end
+
+"""
+Define the matching test for each case in a selector, i.e.
+
+```julia
+Selectors.matcher(::Type{First}, x)  = x == 1
+Selectors.matcher(::Type{Second}, x) = true
+```
+
+Note that the return type must be `Bool`.
+
+To match against multiple cases use the [`Selectors.strict`](@ref) function.
+"""
+function matcher end
+
+"""
+Define the code that will run when a particular [`Selectors.matcher`](@ref) test returns
+`true`, i.e.
+
+```julia
+Selectors.runner(::Type{First}, x)  = println("`x` is equal to `1`.")
+Selectors.runner(::Type{Second}, x) = println("`x` is not equal to `1`.")
+```
+"""
+function runner end
+
+"""
+Define whether a selector case will "fallthrough" or not when successfully matched against.
+By default matching is strict and does not fallthrough to subsequent selector cases.
+
+```julia
+# Adding a debugging selector case.
+abstract Debug <: MySelector
+
+# Insert prior to all other cases.
+Selectors.order(::Type{Debug}) = 0.0
+
+# Fallthrough to the next case on success.
+Selectors.strict(::Type{Debug}) = false
+
+# We always match, regardless of the value of `x`.
+Selectors.matcher(::Type{Debug}, x) = true
+
+# Print some debugging info.
+Selectors.runner(::Type{Debug}, x) = @show x
+```
+"""
+strict{T <: AbstractSelector}(::Type{T}) = true
+
+"""
+Generated function that builds a specialised selector for each selector type provided, i.e.
+
+```julia
+Selectors.selector(MySelector, 1)
+```
+"""
+@generated function selector{T <: AbstractSelector}(::Type{T}, x...)
```

Comment:
Not sure about this, but would it make sense to have the function name be a verb? `select`? Or maybe `dispatch`?